rfcs/require-by-string.md

```diff
@@ -0,0 +1,360 @@
+
+To require a Luau module under the current implementation, we must require it either by relative or absolute path:
+
+```lua
+-- Requiring Roact by absolute path
+local Roact = require("C:/LuauModules/Roact-v1.4.2")
+```
+
+If we wanted to change the location of this package on disk, we would have to modify all files that require the old path above, which essentially boils down to manual package management. This approach is tedious and prone to mistakes.
+
+To solve this, we introduce aliases in this RFC, which would allow for future package management software to maintain package paths and versions by directly modifying alias maps.
+
+This would also create simple and readable require statements for developers.
+
+```lua
+-- Suppose "@src" is an alias for the same directory as "../../../../../"
+
+-- Instead of this:
+local result = require("../../../../../HelperModules/graphing")
+
+-- We could have this:
+local result = require("@src/HelperModules/graphing")
+```
+
+## Design
+
+### String syntax
+
+Our require syntax should:
+
+- Allow a library written in Luau to be imported into the Roblox engine and "just work".
+- Support compile-time file resolution _where possible_ for type checking.
+- Be consistent across platforms and both inside and outside of Roblox.
+- Provide support for binding unambiguous aliases to relative and absolute paths.
+- Attempt to remain consistent with the existing Luau path syntax.
+
+For compatibility across platforms, we will automatically map `/` onto `\`.
+
+#### Relative paths
+
+Modules can be required relative to the requiring file's location in the filesystem (note, this is different from the current implementation, which evaluates all relative paths in relation to the current working directory).
+
+If we are trying to require a module called `MyModule.luau` in `C:/MyLibrary`:
+```lua
+-- From C:/MyLibrary/SomeModule.luau
+local MyModule = require("MyModule")
+
+-- Same as above, identical behavior
+local MyModule = require("./MyModule")
+
+-- From C:/MyLibrary/SubDirectory/SubModule.luau
+local MyModule = require("../MyModule")
+ 
+-- From C:/MyOtherLibrary/MainModule.luau
+local MyModule = require("../MyLibrary/MyModule")
+```
+
+Relative paths are chosen as the default path type, meaning that if a given path does not begin with a reserved prefix such as `/` or `@`, then it is considered relative to the requiring file's location. This is chosen for convenience, as libraries may include various internal dependencies that are easier to reference using relative paths.
+
+Relative paths can also begin with `./` or `../`, which denote the directory of the requiring file and its parent directory, respectively. When executing Luau code directly from a CLI (e.g. in the Luau CLI), this is relative to the current working directory.
+
+#### Absolute paths
+
+Modules can also be required by their absolute path by prefixing the root of the filesystem (e.g. `C:/` or `/`).
```

Comment:
Since the current implementation just uses `fopen`/`_wfopen` (which supports both relative-to-cwd and absolute paths), doing this would break backwards compatibility, even if we are supporting them through aliases.

Reply:
Aren't we guaranteed to break it anyway because the way relative paths behave must change?

Reply:
I suppose everything is alright as long as we provide an alternative (and aren't just deprecating support for absolute paths altogether). To clarify, you're saying we shouldn't support `require("/some/absolute/path")`, but it's okay to support absolute paths by feeding them in through an external source (like `LUAU_PATH` or an alias map defined in `.luaurc`)?

Reply:
Yeah exactly. The way I'm thinking about this is that the syntax inside require() should not ever be tied to a file system. Because it needs to execute on different OSes, users, and also be portable between real filesystems and virtual filesystems (like Roblox DOM), it must be a small subset of total functionality that a given OS provides you. So we should restrict it as much as we can initially. The fact that we have some scripts today that run through Luau CLI that might break as a result I think is fine because I think this is inevitable - I'm in fact more interested in improving compatibility between Lua CLI require and new require, than I am interested in keeping compatibility between current Luau CLI require and new require :) we sort of intentionally shipped a very small minimal require implementation with the goal of eventually making a full fledged version of it, so I don't think we have a lot of cases where the exact behavior is relied upon.